### PR TITLE
PRO-16384: Create VSCode extension for ProvarDX

### PR DESCRIPTION
**Pull request by riteshnahar**
- Added support for ProvarDX commands: Create Properties File, ProvarDX: Validate Properties File, ProvarDX: Run Testcases

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,21 +1,32 @@
 {
     "name": "provardx-vscode",
-    "displayName": "provardx-vscode",
-    "description": "ProvarDX support for VS Code",
+    "displayName": "ProvarDX",
```
Comment (ProvarTesting) [inline package.json:3]: change to provardx-vscode

File: yarn-error.log [hunk 2]
```diff
@@ -0,0 +1,2095 @@
+Arguments:
```
Comment (ProvarTesting) [inline yarn-error.log]: Do we need to commit this file?

File: README.md [hunk 3]
```diff
@@ -1 +1,19 @@
-# provardx-vscode
\ No newline at end of file
+# ProvarDX SFDX Plugin Integration for Visual Studio Code
+
+This extension enables Visual Studiio Code to run ProvarDX CLI commands in an interactive way. Existing Provar customers can execute
+Provar Test Cases from Visual Studio Code. You must be an existing Provar customer with a valid paid license to write and maintain your test cases.
+
+![GIF showing Create properties file, validate and running tests](/images/provardx-preview.gif)
```
Comment (ProvarTesting) [inline README.md:6]: Is it possible to delay the video to see the result of the run test command as it starts the replay of the video quite quickly?

File: tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+        "target": "es6",
+        "outDir": "out",
+        "lib": ["es6"],
+        "sourceMap": true,
+        "rootDir": "src",
+        "strict": true /* enable all strict type-checking options */
+        /* Additional Checks */
+        // "noImplicitReturns": true, /* Report error when not all code paths in function return a value. */
+        // "noFallthroughCasesInSwitch": true, /* Report errors for fallthrough cases in switch statement. */
+        // "noUnusedParameters": true,  /* Report errors on unused parameters. */
```
Comment (ProvarTesting) [inline tsconfig.json]: do we need this commented out code?

File: src/commands/createPropertiesFile.ts [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+            case 'darwin':
+                return path.join('/Applications', 'Provar.app', 'Contents', 'Eclipse');
+            case 'win32':
+                return process.env.PROVAR_HOME || path.join('C:\\', 'Program Files', 'Provar');
+        }
+    }
+}
+
+export default async function createPropertiesFile() {
+    const createPropertiesFile = new CreatePropertiesFile();
```
Comment (ProvarTesting) [inline src/commands/createPropertiesFile.ts]: rename the variable

File: src/commands/runTests.ts [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+async function killPromise(processId: number, signal: string): Promise<void> {
+    return new Promise<void>((resolve, reject) => {
+        kill(processId, signal, (err: {}) => {
+            err ? reject(err) : resolve();
+        });
+    });
+}
+
+export default async function runTests() {
+    const runTests = new RunTests();
```
Comment (ProvarTesting) [inline src/commands/runTests.ts]: rename the variable

File: src/commands/validate.ts [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+            ProgressNotification.show(execution, cancellationTokenSource, COMMANDS.PROVARDX_VALIDATE);
+            channelService.showChannelOutput();
+            channelService.streamCommandOutput(execution);
+        }
+    }
+}
+
+export default async function validate() {
+    const validate = new Validate();
```
Comment (ProvarTesting) [inline src/commands/validate.ts]: rename the variable

File: src/utils/preConditionChecker.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+                    err.message.indexOf(messages.error_mac_sfdx_comand_not_found) > -1 ||
+                    err.message.indexOf(messages.error_win32_sfdx_comand_not_found) > -1
+                        ? vscode.window.showErrorMessage(messages.error_sfdx_plugin_not_installed)
+                        : vscode.window.showErrorMessage(`${messages.error} ${err.message}`);
+                    console.error(`${messages.error} ${err}`);
+                    return reject(false);
+                }
+                const response = stdout.toString().trim();
+                if (response === messages.error_no_plugins_installed || response.indexOf(PLUGIN_NAME) === -1) {
+                    vscode.window.registerUriHandler;
```
Comment (ProvarTesting) [inline src/utils/preConditionChecker.ts]: are we missing something here?

File: src/utils/commandExecutor.ts [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        this.cancellationToken = cancellationToken;
+
+        const childProcess = cross_spawn(this.command.command, this.command.args, {});
+
+        this.childProcessPid = childProcess.pid;
+
+        let timerSubscriber: Subscription | null;
+
+        // Process
+        this.processExitSubject = fromEvent(childProcess, 'exit') as Observable<number | undefined>;
```
Comment (ProvarTesting) [inline src/utils/commandExecutor.ts]: redundant cast as we're not change the type?

File: src/utils/commandExecutor.ts [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        this.processExitSubject.subscribe((next) => {
+            if (timerSubscriber) {
+                timerSubscriber.unsubscribe();
+            }
+        });
+        this.processErrorSubject = fromEvent(childProcess, 'error') as Observable<Error | undefined>;
```
Comment (ProvarTesting) [inline src/utils/commandExecutor.ts]: redundant cast as we're not change the type?

File: src/test/suite/index.ts [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+            try {
+                // Run the mocha test
+                mocha.run((failures) => {
+                    if (failures > 0) {
+                        e(new Error(`${failures} tests failed.`));
+                    } else {
+                        c();
+                    }
+                });
+            } catch (err) {
```
Comment (ProvarTesting) [inline src/test/suite/index.ts]: variable already declared in upper scope, rename the variable

File: package.json [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+    "husky": {
+        "hooks": {
+            "pre-commit": "lint-staged"
+        }
+    },
+    "lint-staged": {
+        "**/*.{json,md,ts}": [
+            "prettier --write"
+        ]
+    }
```
Comment (ProvarTesting) [inline package.json:93]: can we pull in the salesforce extension dependencies using "extensionPack": [...

File: README.md [hunk 13]
```diff
@@ -1 +1,19 @@
...
+
+## Prerequisites
```
Comment (ProvarTesting) [inline README.md:8]: add sections for:
- Documentation, we'll need to add a link to the Provar support pages
- Bugs and feedback
- Included extensions

File: README.md [hunk 14]
```diff
@@ -1 +1,19 @@
...
+
+Before you set up ProvarDX CLI Integration for VS Code, make sure that you have these essentials.
+
+-   **Provar**
```
Comment (ProvarTesting) [inline README.md:12]: Do we need both Provar Desktop and ProvarDX installed to use this?

**Review by ProvarTesting — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (riteshnahar) [reply package.json:3]: Changing the display name will change the name of the plugin in VSCode (attached screenshot).
Screenshot 2020-07-18 at 12 28 10 PM

File: README.md (hunk 3 above)
Comment (riteshnahar) [reply README.md:6]: I have explicitly omitted that scenario as the gif size will increase if we show the run test command output.
Still, if you say we can try it.

File: tsconfig.json (hunk 4 above)
Comment (riteshnahar) [reply tsconfig.json]: Removed

File: src/utils/preConditionChecker.ts (hunk 8 above)
Comment (riteshnahar) [reply src/utils/preConditionChecker.ts]: Removed this line.

File: yarn-error.log (hunk 2 above)
Comment (riteshnahar) [reply yarn-error.log]: Removed

File: src/commands/createPropertiesFile.ts (hunk 5 above)
Comment (riteshnahar) [reply src/commands/createPropertiesFile.ts]: done

File: src/commands/runTests.ts (hunk 6 above)
Comment (riteshnahar) [reply src/commands/runTests.ts]: done

File: src/commands/validate.ts (hunk 7 above)
Comment (riteshnahar) [reply src/commands/validate.ts]: done

File: src/utils/commandExecutor.ts (hunk 9 above)
Comment (riteshnahar) [reply src/utils/commandExecutor.ts]: done

File: src/utils/commandExecutor.ts (hunk 10 above)
Comment (riteshnahar) [reply src/utils/commandExecutor.ts]: done

File: src/test/suite/index.ts (hunk 11 above)
Comment (riteshnahar) [reply src/test/suite/index.ts]: done

File: package.json (hunk 12 above)
Comment (riteshnahar) [reply package.json:93]: Nope, we do not require Salesforce Extensions in our case.

File: README.md (hunk 14 above)
Comment (riteshnahar) [reply README.md:12]: Yes we would need both.

File: README.md (hunk 13 above)
Comment (riteshnahar) [reply README.md:8]: done

File: README.md (hunk 3 above)
Comment (ProvarTesting) [reply README.md:6]: Could you try it as the video does start the replay quite quickly and can't really see what happened. It'll be on the market place so worth showing it's full functionality if possible, just make a really short test.

**Review by ProvarTesting — APPROVED** [review]

**Pull request merged**